src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = state;
+
+  actions.forEach((action) => {
+    let nextState;
+
+    switch (action.type) {
+      case 'clear':
+        nextState = {};
+        break;
+      case 'addProperties':
+        nextState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        nextState = { ...currentState };
+
+        action.keysToRemove.forEach((key) => {
+          if (nextState.hasOwnProperty(key)) {
```

Comment:
Using 'hasOwnProperty' directly from the object is not recommended due to potential issues with properties that may shadow prototype properties. Instead, use 'Object.prototype.hasOwnProperty.call(nextState, key)' to avoid such issues.